### PR TITLE
GetRubric endpoint

### PR DESCRIPTION
**Pull request by EAndrey1091**
Added GetRubric endpoint
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/1042

### CODE REVIEW

File: src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+namespace LT.DigitalOffice.WikiService.Business.Commands.Rubric
+{
+  public class GetRubricCommand : IGetRubricCommand
+  {
+    private readonly IRubricRepository _rubricRepository;
+    private readonly IRubricResponseMapper _rubricResponseMapper;
+    private readonly IHttpContextAccessor _httpContextAccessor;
+    private readonly IResponseCreator _responseCreator;
+
```
Comment (MariyaSin) [inline src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs:19]: remove empty line

File: src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      IRubricRepository rubricRepository,
+      IRubricResponseMapper rubricResponseMapper,
+      IHttpContextAccessor httpContextAccessor,
+      IResponseCreator responseCreator)
+    {
+      _rubricRepository = rubricRepository;
+      _rubricResponseMapper = rubricResponseMapper;
+      _httpContextAccessor = httpContextAccessor;
+      _responseCreator = responseCreator;
+    }
```
Comment (MariyaSin) [inline src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs:28]: add empty line

File: src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    public async Task<OperationResultResponse<RubricResponse>> ExecuteAsync(GetRubricFilter filter)
+    {
+      OperationResultResponse<RubricResponse> response = new();
+
+      if (filter is null ||
+        filter.RubricId is null)
+      {
+        return _responseCreator.CreateFailureResponse<RubricResponse>(
+          HttpStatusCode.BadRequest,
+          new List<string> { "You must specify 'rubricId'." });
```
Comment (MariyaSin) [inline src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs]: rubricId can not be null in get request

File: src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+      }
+
+      List<DbRubric> subRubrics = await _rubricRepository.GetSubRubricsAsync(filter);
+
+      DbRubric dbRubric = await _rubricRepository.GetAsync(filter);
+
+      response.Body = _rubricResponseMapper.Map(
+        dbRubric, subRubrics);
+
+      response.Status = OperationResultStatusType.FullSuccess;
```
Comment (MariyaSin) [inline src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs]: sets by default

File: src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+      if (response.Body == null)
+      {
+        _httpContextAccessor.HttpContext.Response.StatusCode = (int)System.Net.HttpStatusCode.NotFound;
```
Comment (MariyaSin) [inline src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs]: use response creator

File: src/WikiService.Models.Dto/Responses/Rubric/RubricResponse.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
+﻿using LT.DigitalOffice.WikiService.Models.Dto.Models;
+using System;
+using System.Collections.Generic;
+
```
Comment (MariyaSin) [inline src/WikiService.Models.Dto/Responses/Rubric/RubricResponse.cs:3]: remove empty line

File: src/WikiService.Models.Dto/Responses/Rubric/RubricResponse.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace LT.DigitalOffice.WikiService.Models.Dto.Responses.Rubric
+{
+  public class RubricResponse
```
Comment (MariyaSin) [inline src/WikiService.Models.Dto/Responses/Rubric/RubricResponse.cs]: record

File: src/WikiService.Models.Dto/Requests/Rubric/Filters/GetRubricFilter.cs [hunk 8]
```diff
@@ -0,0 +1,18 @@
+﻿using LT.DigitalOffice.Kernel.Requests;
+using Microsoft.AspNetCore.Mvc;
+using System;
+
+namespace LT.DigitalOffice.WikiService.Models.Dto.Requests.Rubric.Filters
+{
+  public record GetRubricFilter : BaseFindFilter
```
Comment (MariyaSin) [inline src/WikiService.Models.Dto/Requests/Rubric/Filters/GetRubricFilter.cs]: it is not find endpoint

File: src/WikiService.Models.Dto/Requests/Rubric/Filters/GetRubricFilter.cs [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+  {
+    [FromQuery(Name = "rubricId")]
+    public Guid? RubricId { get; set; }
```
Comment (MariyaSin) [inline src/WikiService.Models.Dto/Requests/Rubric/Filters/GetRubricFilter.cs]: get request can not have nullable the entity id

File: src/WikiService.Models.Dto/Requests/Rubric/Filters/GetRubricFilter.cs [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+    [FromQuery(Name = "includeSubRubrics")]
+    public bool IncludeSubRubrics { get; set; }
```
Comment (MariyaSin) [inline src/WikiService.Models.Dto/Requests/Rubric/Filters/GetRubricFilter.cs]: set flse by default

File: src/WikiService.Models.Dto/Requests/Rubric/Filters/GetRubricFilter.cs [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+    [FromQuery(Name = "includeArticles")]
+    public bool IncludeArticles { get; set; }
```
Comment (MariyaSin) [inline src/WikiService.Models.Dto/Requests/Rubric/Filters/GetRubricFilter.cs]: set false by default

File: src/WikiService.Models.Dto/Models/SubRubricInfo.cs [hunk 12]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace LT.DigitalOffice.WikiService.Models.Dto.Models
+{
+  public record SubRubricInfo
```
Comment (MariyaSin) [inline src/WikiService.Models.Dto/Models/SubRubricInfo.cs]: RubricInfo

File: src/WikiService.Models.Dto/Models/SubRubricInfo.cs [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+  {
+    public Guid SubRubricId { get; set; }
```
Comment (MariyaSin) [inline src/WikiService.Models.Dto/Models/SubRubricInfo.cs]: Id

File: src/WikiService.Models.Dto/Models/SubRubricInfo.cs [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+    public string SubRubricName { get; set; }
```
Comment (MariyaSin) [inline src/WikiService.Models.Dto/Models/SubRubricInfo.cs]: Name

File: src/WikiService.Mappers/Responses/RubricResponseMapper.cs [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+      return new RubricResponse
+      {
+        Id = dbRubric.Id,
+        Name = dbRubric.Name,
+        ParentId = dbRubric.ParentId,
+        IsActive = dbRubric.IsActive,
+
+        SubRubrics = subRubrics
+        ?.Select(x => new SubRubricInfo
```
Comment (MariyaSin) [inline src/WikiService.Mappers/Responses/RubricResponseMapper.cs]: add tab

File: src/WikiService.Models.Dto/Models/ArticleInfo.cs [hunk 16]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+
+namespace LT.DigitalOffice.WikiService.Models.Dto.Models
+{
+  public record ArticleInfo
+  {
+    public Guid? ArticleId { get; set; }
```
Comment (MariyaSin) [inline src/WikiService.Models.Dto/Models/ArticleInfo.cs]: Article id can not be null!!!

Comment (MariyaSin) [reply src/WikiService.Models.Dto/Models/ArticleInfo.cs]: rename to Id

File: src/WikiService.Models.Dto/Models/ArticleInfo.cs [hunk 17]
```diff
@@ -0,0 +1,10 @@
...
+    public string ArticleName { get; set; }
```
Comment (MariyaSin) [inline src/WikiService.Models.Dto/Models/ArticleInfo.cs]: rename to Name

File: src/WikiService.Mappers/Responses/RubricResponseMapper.cs [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace LT.DigitalOffice.WikiService.Mappers.Responses
+{
+  public class RubricResponseMapper : IRubricResponseMapper
+  {
+    public RubricResponse Map(
+      DbRubric dbRubric,
+      List<DbRubric> subRubrics)
+    {
+
```
Comment (MariyaSin) [inline src/WikiService.Mappers/Responses/RubricResponseMapper.cs:22]: remove empty line

File: src/WikiService.Mappers/Responses/RubricResponseMapper.cs [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+      if (dbRubric == null)
```
Comment (MariyaSin) [inline src/WikiService.Mappers/Responses/RubricResponseMapper.cs]: use "is"

File: src/WikiService.Data/RubricRepository.cs [hunk 20]
```diff
@@ -112,5 +112,32 @@ public async Task<bool> DoesRubricNameExistAsync(Guid? rubricParentId, string na
     {
       return await _provider.Rubrics.AnyAsync(p => p.ParentId == rubricParentId && p.Name == nameRubric);
     }
+
+    public async Task<List<DbRubric>> GetSubRubricsAsync(GetRubricFilter filter)
```
Comment (MariyaSin) [inline src/WikiService.Data/RubricRepository.cs]: why are you creating a separate method? move it to the "get" method

File: src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs (hunk 5 above)
Comment (EAndrey1091) [reply src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs]: Done

File: src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs (hunk 4 above)
Comment (EAndrey1091) [reply src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs]: Done

File: src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs (hunk 3 above)
Comment (EAndrey1091) [reply src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs]: Done

File: src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs (hunk 2 above)
Comment (EAndrey1091) [reply src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs:28]: Done

File: src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs (hunk 1 above)
Comment (EAndrey1091) [reply src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs:19]: Done

File: src/WikiService.Models.Dto/Responses/Rubric/RubricResponse.cs (hunk 6 above)
Comment (EAndrey1091) [reply src/WikiService.Models.Dto/Responses/Rubric/RubricResponse.cs:3]: Done

File: src/WikiService.Models.Dto/Responses/Rubric/RubricResponse.cs (hunk 7 above)
Comment (EAndrey1091) [reply src/WikiService.Models.Dto/Responses/Rubric/RubricResponse.cs]: Done

File: src/WikiService.Models.Dto/Requests/Rubric/Filters/GetRubricFilter.cs (hunk 8 above)
Comment (EAndrey1091) [reply src/WikiService.Models.Dto/Requests/Rubric/Filters/GetRubricFilter.cs]: Done

File: src/WikiService.Models.Dto/Requests/Rubric/Filters/GetRubricFilter.cs (hunk 9 above)
Comment (EAndrey1091) [reply src/WikiService.Models.Dto/Requests/Rubric/Filters/GetRubricFilter.cs]: Done

File: src/WikiService.Models.Dto/Requests/Rubric/Filters/GetRubricFilter.cs (hunk 11 above)
Comment (EAndrey1091) [reply src/WikiService.Models.Dto/Requests/Rubric/Filters/GetRubricFilter.cs]: Done

File: src/WikiService.Models.Dto/Models/SubRubricInfo.cs (hunk 12 above)
Comment (EAndrey1091) [reply src/WikiService.Models.Dto/Models/SubRubricInfo.cs]: Done

File: src/WikiService.Models.Dto/Models/SubRubricInfo.cs (hunk 13 above)
Comment (EAndrey1091) [reply src/WikiService.Models.Dto/Models/SubRubricInfo.cs]: Done

File: src/WikiService.Models.Dto/Models/SubRubricInfo.cs (hunk 14 above)
Comment (EAndrey1091) [reply src/WikiService.Models.Dto/Models/SubRubricInfo.cs]: Done

File: src/WikiService.Mappers/Responses/RubricResponseMapper.cs (hunk 15 above)
Comment (EAndrey1091) [reply src/WikiService.Mappers/Responses/RubricResponseMapper.cs]: Done

File: src/WikiService.Models.Dto/Models/ArticleInfo.cs (hunk 16 above)
Comment (EAndrey1091) [reply src/WikiService.Models.Dto/Models/ArticleInfo.cs]: Done

File: src/WikiService.Models.Dto/Models/ArticleInfo.cs (hunk 17 above)
Comment (EAndrey1091) [reply src/WikiService.Models.Dto/Models/ArticleInfo.cs]: Done

File: src/WikiService.Mappers/Responses/RubricResponseMapper.cs (hunk 18 above)
Comment (EAndrey1091) [reply src/WikiService.Mappers/Responses/RubricResponseMapper.cs:22]: Done

File: src/WikiService.Mappers/Responses/RubricResponseMapper.cs (hunk 19 above)
Comment (EAndrey1091) [reply src/WikiService.Mappers/Responses/RubricResponseMapper.cs]: Done

File: src/WikiService.Data/RubricRepository.cs (hunk 20 above)
Comment (EAndrey1091) [reply src/WikiService.Data/RubricRepository.cs]: Done

File: src/WikiService.Models.Dto/Requests/Rubric/Filters/GetRubricFilter.cs (hunk 10 above)
Comment (EAndrey1091) [reply src/WikiService.Models.Dto/Requests/Rubric/Filters/GetRubricFilter.cs]: Done

File: src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+      _responseCreator = responseCreator;
+    }
+
+    public async Task<OperationResultResponse<RubricResponse>> ExecuteAsync(GetRubricFilter filter)
+    {
+      DbRubric dbRubric = await _rubricRepository.GetAsync(filter);
+      
+      OperationResultResponse<RubricResponse> response = new();
+
+      response.Body = _rubricResponseMapper.Map(dbRubric);
```
Comment (MariyaSin) [inline src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs]: join with 37 line

File: src/WikiService.Mappers/Responses/RubricResponseMapper.cs (hunk 18 above)
Comment (MariyaSin) [reply src/WikiService.Mappers/Responses/RubricResponseMapper.cs:22]: not removed

File: src/WikiService.Mappers/Responses/RubricResponseMapper.cs [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+          ?.Select(x => new RubricInfo
+          {
+            Id = x.Id,
+            Name = x.Name,
+            ParentId = x.ParentId,
+            IsActive = x.IsActive,
+            HasChild = x.HasChild
+      }),
+
+        Articles = dbRubric.Articles
```
Comment (MariyaSin) [inline src/WikiService.Mappers/Responses/RubricResponseMapper.cs]: use articleInfoMapper

File: src/WikiService.Mappers/Responses/RubricResponseMapper.cs [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+
+      return new RubricResponse
+      {
+        Id = dbRubric.Id,
+        Name = dbRubric.Name,
+        ParentId = dbRubric.ParentId,
+        IsActive = dbRubric.IsActive,
+
+        SubRubrics = dbRubric.ParentIds
+          ?.Select(x => new RubricInfo
```
Comment (MariyaSin) [inline src/WikiService.Mappers/Responses/RubricResponseMapper.cs]: use rubricInfoMapper

File: src/WikiService.Mappers/Responses/RubricResponseMapper.cs [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+        return null;
+      }
+
+      return new RubricResponse
+      {
+        Id = dbRubric.Id,
+        Name = dbRubric.Name,
+        ParentId = dbRubric.ParentId,
+        IsActive = dbRubric.IsActive,
+
```
Comment (MariyaSin) [inline src/WikiService.Mappers/Responses/RubricResponseMapper.cs:43]: remove empy line

File: src/WikiService.Mappers/Responses/RubricResponseMapper.cs [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+        SubRubrics = dbRubric.ParentIds
+          ?.Select(x => new RubricInfo
+          {
+            Id = x.Id,
+            Name = x.Name,
+            ParentId = x.ParentId,
+            IsActive = x.IsActive,
+            HasChild = x.HasChild
+      }),
+
```
Comment (MariyaSin) [inline src/WikiService.Mappers/Responses/RubricResponseMapper.cs:48]: remove empty line

File: src/WikiService.Models.Db/DbRubric.cs [hunk 26]
```diff
@@ -18,9 +18,11 @@ public class DbRubric
     public DateTime CreatedAtUtc { get; set; }
     public Guid? ModifiedBy { get; set; }
     public DateTime? ModifiedAtUtc { get; set; }
-
+
```
Comment (MariyaSin) [inline src/WikiService.Models.Db/DbRubric.cs]: remove tabs

File: src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs (hunk 21 above)
Comment (EAndrey1091) [reply src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs]: Done

File: src/WikiService.Mappers/Responses/RubricResponseMapper.cs (hunk 23 above)
Comment (EAndrey1091) [reply src/WikiService.Mappers/Responses/RubricResponseMapper.cs]: Done

File: src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    public async Task<OperationResultResponse<RubricResponse>> ExecuteAsync(GetRubricFilter filter)
+    {
+      OperationResultResponse<RubricResponse> response = new();
+
+      response.Body = 
+        _rubricResponseMapper.Map(await _rubricRepository.GetAsync(filter));
+
+      return response.Body == null
```
Comment (MariyaSin) [inline src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs]: please, use "is" instead "=="

File: src/WikiService.Mappers/Responses/RubricResponseMapper.cs [hunk 28]
```diff
@@ -0,0 +1,40 @@
...
+      {
+        return null;
+      }
+
+      return new RubricResponse
+      {
+        Rubric = _rubricInfoMapper.Map(dbRubric),
+
+        SubRubrics = dbRubric.SubRubrics
+        ?.Select(x => _rubricInfoMapper.Map(x)),
```
Comment (MariyaSin) [inline src/WikiService.Mappers/Responses/RubricResponseMapper.cs]: add tab

File: src/WikiService.Mappers/Responses/RubricResponseMapper.cs [hunk 29]
```diff
@@ -0,0 +1,40 @@
...
+
+        Articles = dbRubric.Articles
+        ?.Select(x => _articleInfoMapper.Map(x))
```
Comment (MariyaSin) [inline src/WikiService.Mappers/Responses/RubricResponseMapper.cs]: add tab

File: src/WikiService.Mappers/Responses/RubricResponseMapper.cs [hunk 30]
```diff
@@ -0,0 +1,40 @@
...
+  {
+    private readonly IRubricInfoMapper _rubricInfoMapper;
+    private readonly IArticleInfoMapper _articleInfoMapper;
+    public RubricResponseMapper(
+      IRubricInfoMapper rubricInfoMapper,
+      IArticleInfoMapper articleInfoMapper)
+    {
+      _rubricInfoMapper = rubricInfoMapper;
+      _articleInfoMapper = articleInfoMapper;
+    }
```
Comment (MariyaSin) [inline src/WikiService.Mappers/Responses/RubricResponseMapper.cs:21]: add empty line

File: src/WikiService.Mappers/Responses/RubricResponseMapper.cs [hunk 31]
```diff
@@ -0,0 +1,40 @@
...
+using LT.DigitalOffice.WikiService.Models.Db;
+using LT.DigitalOffice.WikiService.Models.Dto.Responses.Rubric;
+using System.Linq;
+
+namespace LT.DigitalOffice.WikiService.Mappers.Responses
+{
+  public class RubricResponseMapper : IRubricResponseMapper
+  {
+    private readonly IRubricInfoMapper _rubricInfoMapper;
+    private readonly IArticleInfoMapper _articleInfoMapper;
```
Comment (MariyaSin) [inline src/WikiService.Mappers/Responses/RubricResponseMapper.cs:13]: add empty line

File: src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs [hunk 32]
```diff
@@ -0,0 +1,44 @@
...
+      OperationResultResponse<RubricResponse> response = new();
+
+      (DbRubric dbRubric, filter.IncludeSubRubrics) = await _rubricRepository.GetAsync(filter);
+
+      response.Body = 
+        _rubricResponseMapper.Map(dbRubric, filter.IncludeSubRubrics);
+
+      return response.Body is null
+        ? _responseCreator.CreateFailureResponse<RubricResponse>(HttpStatusCode.NotFound)
+:       response;
```
Comment (vorobevamv) [inline src/WikiService.Business/Commands/Rubric/GetRubricCommand.cs]: fix space (between ":" and "response") and tab

**Pull request closed without merging**